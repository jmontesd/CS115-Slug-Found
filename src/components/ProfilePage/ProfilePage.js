import React from 'react';
import './ProfilePage.scss';
import { compose } from 'redux';
import { firestoreConnect } from 'react-redux-firebase';
import { connect } from 'react-redux';
import { Redirect, Link } from 'react-router-dom';
import Item from '../Item/Item';
import { updateProfileImage as updateProfileImageAction } from '../../store/actions/authActions';

export const ProfilePage = (props) => {
  // these are the props need for this component
  const {
    id,
    username,
    posts,
    isLoggedIn,
    profilePictureURL,
    isUserProfileTheUserLoggedIn,
    updateProfileImage,
  } = props;
  // redirect use to login if they are not logged in
  if (!isLoggedIn) return <Redirect to="/login" />;
  // this is used to handle when user uploads image
  const handleFileChange = (e) => {
    // check to make sure they submitted file
    if (e.target.files[0]) {
      const imageFile = e.target.files[0];
      // update their profile image
      updateProfileImage(imageFile);
    }
  };
  // render this to the screen
  return (
    <div className="container">
      <div className="row">
        <div className="col-8">
          {/* if there are posts, render them to the screen */}
          {posts &&
            posts
              .slice()
              .sort((a, b) => b.createdAt - a.createdAt)
              .map((post) => <Item key={post.id} post={post} />)}
        </div>
        <div className="col-4">
          <div className="profile-wrapper">
            <img src={profilePictureURL} className="img-card" alt="Profile" />
            {/* if the user profile is the user logged in, render change profile picture button */}
            {isUserProfileTheUserLoggedIn && (
              <label className="btn btn-outline-info cursor-pointer d-block mt-2 ">
                <input className="d-none" type="file" onChange={handleFileChange} />
                Change Profile Picture
              </label>
            )}
            <div className="card-title">{username}</div>
            <Link to={`/messages/message/${id}`} className="btn btn-warning full-width">
              Message
            </Link>
          </div>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = (state, props) => {
  // get user id of the profile currently visiting
  const { id } = props.match.params;
  // get all posts and all users
  const { posts: allPosts, users } = state.firestore.ordered;
  // find the user
  const user = users && users.find((i) => i.id === id);
  // get their username
  const username = user && user.username;
  // get their picture
  const profilePictureURL = user && user.profilePictureURL;
  // find there posts
  const posts = allPosts && allPosts.filter((i) => i.user.id === id);
  // check if user profile is the user logged in
  const isUserProfileTheUserLoggedIn = id === state.firebase.auth.uid;
  return {
    id,
    isLoggedIn: state.firebase.auth.uid,
    isUserProfileTheUserLoggedIn,
    profilePictureURL,
    posts,
    username,
  };
};
// get the updateProfileImage function which will update profile image
// for the user
const mapDispatchToProps = (dispatch) => ({
  updateProfileImage: (update) => dispatch(updateProfileImageAction(update)),
});
// export this component with the neccessary data
export default compose(
  connect(
    mapStateToProps,
    mapDispatchToProps,
  ),
  // connect firestore with all posts and users
  firestoreConnect([{ collection: 'posts' }, { collection: 'users' }]),
)(ProfilePage);
