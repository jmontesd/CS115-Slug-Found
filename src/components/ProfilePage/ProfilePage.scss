.profile-wrapper {
  border: 1px solid rgb(200, 200, 200);
  border-radius: 0.25rem;
  padding: 1rem;
}

.card-title {
  font-size: 1.5rem;
}

.img-card {
  height: auto;
  width: auto;
  max-height: 100px;
}

.full-width {
  width: 100%;
}
