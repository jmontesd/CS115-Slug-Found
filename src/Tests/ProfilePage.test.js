import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { ProfilePage } from '../components/ProfilePage/ProfilePage';

// needed for shallow rendering
configure({ adapter: new Adapter() });

describe('<ProfilePage />', () => {
  it('should render ProfilePage properly', () => {
    // test creates snapshot of component and checks if matches
    // previously create snapshot
    const wrapper = shallow(<ProfilePage isLoggedIn username="username" />);
    expect(wrapper).toMatchSnapshot();
  });
  it('should render updateProfileImage properly', () => {
    // mock function
    const updateProfileImage = jest.fn();
    const wrapper = shallow(
      <ProfilePage
        isLoggedIn
        isUserProfileTheUserLoggedIn
        updateProfileImage={updateProfileImage}
      />,
    );
    // simulate adding file
    wrapper.find('input').simulate('change', { target: { files: [1, 2] } });
    // check if mock function was called
    expect(updateProfileImage).toHaveBeenCalled();
  });
});
