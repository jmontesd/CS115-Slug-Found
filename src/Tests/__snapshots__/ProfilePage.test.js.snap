// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProfilePage /> should render ProfilePage properly 1`] = `ShallowWrapper {}`;
